pypeit/images/rawimage.py

```diff
@@ -424,38 +726,100 @@ def subtract_pattern(self):
+        # order of forcing the calculation of the readnoise directly after
+        # subtracting the pattern get around this?
+        # TODO: Since subtract_pattern effectively must happen first, does it
+        # make sense to pass self.rawimage instead of self.image?
+        # TODO: Is there an error estimate in the pattern subtraction that we
+        # can propagate to the error budget?
+        self.rawimage = procimg.subtract_pattern(self.image, self.datasec_img, self.oscansec_img,
+                                                 frequency=frequency)
+        self.image = self.rawimage.copy()
         self.steps[step] = True
-        self.image = temp.copy()
-        self.rawimage = temp.copy()
 
     def trim(self, force=False):
         """
-        Trim the image to include only the science data
+        Trim :attr:`image`, :attr:`var` (if it exists), :attr:`rn2img` (if it
+        exists), and :attr:`datasec_img` to include only the science data.
 
         Args:
-            force (bool, optional):
-                Force the processing even if the image was already processed
-
+            force (:obj:`bool`, optional):
+                Force the image to be trimmed, even if the step log
+                (:attr:`steps`) indicates that it already has been.
         """
         step = inspect.stack()[0][3]
-        # Check input image matches the original
         if self.rawimage.shape is not None and self.image.shape != self.rawimage.shape:
-            msgs.warn("Image shape does not match original.  Returning current image")
-            return self.image.copy()
-        # Check if already trimmed
-        if self.steps[step] and (not force):
-            msgs.warn("Image was already trimmed.  Returning current image")
-            return self.image
-        # Do it
+            # Image *must* have been trimmed already because shape does not
+            # match raw image
+            self.steps[step] = True
+            msgs.warn('Image shape does not match original.')
+            return
+        if self.steps[step] and not force:
+            # Already trimmed
+            msgs.warn('Image was already trimmed.')
+            return
         self.image = procimg.trim_frame(self.image, self.datasec_img < 1)
+        if self.var is not None:
+            self.var = procimg.trim_frame(self.var, self.datasec_img < 1)
+        if self.rn2img is not None:
+            self.rn2img = procimg.trim_frame(self.rn2img, self.datasec_img < 1)
+        # NOTE: This must be done last because the untrimmed image is used for
+        # deciding what to trim!
         self.datasec_img = procimg.trim_frame(self.datasec_img, self.datasec_img < 1)
-        #
         self.steps[step] = True
 
+    def add_shot_noise(self):
+        r"""
+        Use the image to calculate and add shot noise to the error budget.
+
+        The method assumes the image is bias- and dark-subtracted, and that the
```

Comment:
Add overscan to this list. Note sometimes we will only overscan subtract, sometimes only overscan subtract and bias subtract, other times only dark-subtract.

Reply:
This code is now different.